source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2052 @@
+static int s_socket_write_fn(
+    struct aws_socket *socket,
+    const struct aws_byte_cursor *cursor,
+    aws_socket_on_write_completed_fn *written_fn,
+    void *user_data);
+static int s_socket_get_error_fn(struct aws_socket *socket);
+static bool s_socket_is_open_fn(struct aws_socket *socket);
+static int s_set_close_callback(struct aws_socket *socket, aws_socket_on_shutdown_complete_fn fn, void *user_data);
+static int s_set_cleanup_callback(struct aws_socket *socket, aws_socket_on_shutdown_complete_fn fn, void *user_data);
+
+static struct aws_socket_vtable s_vtable = {
+    .socket_cleanup_fn = s_socket_cleanup_fn,
+    .socket_connect_fn = s_socket_connect_fn,
+    .socket_bind_fn = s_socket_bind_fn,
+    .socket_listen_fn = s_socket_listen_fn,
+    .socket_start_accept_fn = s_socket_start_accept_fn,
+    .socket_stop_accept_fn = s_socket_stop_accept_fn,
+    .socket_close_fn = s_socket_close_fn,
+    .socket_shutdown_dir_fn = s_socket_shutdown_dir_fn,
+    .socket_set_options_fn = s_socket_set_options_fn,
+    .socket_assign_to_event_loop_fn = s_socket_assign_to_event_loop_fn,
+    .socket_subscribe_to_readable_events_fn = s_socket_subscribe_to_readable_events_fn,
+    .socket_read_fn = s_socket_read_fn,
+    .socket_write_fn = s_socket_write_fn,
+    .socket_get_error_fn = s_socket_get_error_fn,
+    .socket_is_open_fn = s_socket_is_open_fn,
+    .socket_set_close_callback = s_set_close_callback,
+    .socket_set_cleanup_callback = s_set_cleanup_callback,
+};
+
+static void s_schedule_next_read(struct nw_socket *socket);
+
+static void s_socket_cleanup_fn(struct aws_socket *socket) {
+    if (!socket->impl) {
+        /* protect from double clean */
+        return;
+    }
+
+    AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "id=%p nw_socket=%p: is cleanup...", (void *)socket, (void *)socket->impl);
+    if (aws_socket_is_open(socket)) {
+        AWS_LOGF_DEBUG(
+            AWS_LS_IO_SOCKET, "id=%p nw_socket=%p: is still open, closing...", (void *)socket, (void *)socket->impl);
+        aws_socket_close(socket);
+    }
+
+    struct nw_socket *nw_socket = socket->impl;
+
+    if (s_validate_event_loop(socket->event_loop) && !aws_event_loop_thread_is_callers_thread(socket->event_loop)) {
+        s_lock_socket_synced_data(nw_socket);
+        nw_socket->synced_data.base_socket = NULL;
+        s_unlock_socket_synced_data(nw_socket);
+    } else {
+        // If we are already on event loop or event loop is unavailable, we should already acquire the lock for base
+        // socket access
+        nw_socket->synced_data.base_socket = NULL;
+    }
+
+    aws_ref_count_release(&nw_socket->external_ref_count);
+    socket->impl = NULL;
+    AWS_ZERO_STRUCT(*socket);
+}
+
+struct read_queue_node {
+    struct aws_allocator *allocator;
+    dispatch_data_t received_data;
+    struct aws_linked_list_node node;
+    size_t current_offset;
+};
+
+static void s_destroy_read_queue_node(struct read_queue_node *node) {
```

Comment:
Trivial: naming convention on destroy functions. This should be renamed to `s_read_queue_node_destroy()`. Also, if we have a destroy function for `read_queue_node` we should probably set up its pair of `s_read_queue_node_new()` that is called when we are creating something that needs to be destroyed. It looks like there's only one place where this node is created but it feels right to have them both.